docs/sql-distributed-sql-engine-spark-sql-cli.md

```diff
@@ -0,0 +1,192 @@
+        <br>
+        SELECT 1;
+    </code>
+  </td>
+</tr>
+</table>
+
+## Spark SQL CLI Interactive Shell Commands
+
+When `./bin/spark-sql` is run without either the `-e` or `-f` option, it enters interactive shell mode.
+Use `;` (semicolon) to terminate commands. Notice:
+1. The CLI use `;` to terminate commands only when it's at the end of line, and it's not escaped by `\\;`.
+2. `;` is the only way to terminate commands. If the user types `SELECT 1` and presses enter, the console will just wait for input.
+3. If the user types multiple commands in one line like `SELECT 1; SELECT 2;`, the commands `SELECT 1` and `SELECT 2` will be executed separatly.
+4. If `;` appears within a SQL statement (not the end of the line), then it has no special meanings:
+   ```sql
+   -- This is a ; comment
+   SELECT ';' as a;
+   ```
+   This is just a comment line followed by a SQL query which returns a string literal.
+   ```sql
+   /* This is a comment contains ;
+   */ SELECT 1;
+   ```
+   However, if ';' is the end of the line, it terminates the SQL statement. The example above will be terminated into  `/* This is a comment contains ` and `*/ SELECT 1`, Spark will submit these two command and throw parser error (unclosed bracketed comment).
+
+
+<table class="table">
+<tr><th>Command</th><th>Description</th></tr>
+<tr>
+  <td><code>quit</code> or <code>exit</code></td>
+  <td>Exits the interactive shell.</td>
+</tr>
+<tr>
+  <td><code>!&lt;command&gt;</code></td>
+  <td>Executes a shell command from the Spark SQL CLI shell.</td>
+</tr>
+<tr>
+  <td><code>dfs &lt;HDFS dfs command&gt;</code></td>
+  <td>Executes a HDFS <a href="https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HDFSCommands.html#dfs">dfs command</a> from the Spark SQL CLI shell.</td>
+</tr>
+<tr>
+  <td><code>&lt;query string&gt;</code></td>
+  <td>Executes a Spark SQL query and prints results to standard output.</td>
+</tr>
+<tr>
+  <td><code>source &lt;filepath&gt;</code></td>
+  <td>Executes a script file inside the CLI.</td>
+</tr>
+</table>
+
+## Examples
+
+Example of running a query from the command line:
+
+    ./bin/spark-sql -e 'SELECT COL FROM TBL'
+
+Example of setting Hive configuration variables:
+
+    ./bin/spark-sql -e 'SELECT COL FROM TBL' --hiveconf hive.exec.scratchdir=/home/my/hive_scratch
+    
+Example of setting Hive configuration variables:
+
+    ./bin/spark-sql -e 'SELECT ${hiveconf:aaa}' --hiveconf aaa=bbb --hiveconf hive.exec.scratchdir=/home/my/hive_scratch
+    spark-sql> SELECT ${aaa};
+    bbb
+    
+Example of setting Hive variables substitution:
+    
+    ./bin/spark-sql --hivevar aaa=bbb --define ccc=ddd
+    spark-sql> SELECT ${aaa}, ${ccc};
+    bbb ddd
+
+Example of dumping data out from a query into a file using silent mode:
+
+    ./bin/spark-sql -S -e 'SELECT COL FROM TBL' > result.txt
+
+Example of running a script non-interactively from local disk:
+
+    ./bin/spark-sql -f /path/to/spark-sql-script.sql
```

Comment:
does it always mean local file? can `hadoop-site.xml` change the default file scheme?

Reply:
> does it always mean local file? can `hadoop-site.xml` change the default file scheme?

From the code, it can read hdfs file

```
  public int processFile(String fileName) throws IOException {
    Path path = new Path(fileName);
    Object fs;
    if (!path.toUri().isAbsolute()) {
      fs = FileSystem.getLocal(this.conf);
      path = ((FileSystem)fs).makeQualified(path);
    } else {
      fs = FileSystem.get(path.toUri(), this.conf);
    }

    BufferedReader bufferReader = null;
    boolean var5 = false;

    int rc;
    try {
      bufferReader = new BufferedReader(new InputStreamReader(((FileSystem)fs).open(path)));
      rc = this.processReader(bufferReader);
    } finally {
      IOUtils.closeStream(bufferReader);
    }

    return rc;
  }
```

Reply:
then maybe let's remove this example, and add one more example below
```
    ./bin/spark-sql -f file:/path/to/spark-sql-script.sql
    ./bin/spark-sql -f hdfs://<namenode>:<port>/spark-sql-script.sql
    ./bin/spark-sql -f s3://mys3bucket/spark-sql-script.sql
```

Reply:
please verify it locally first. I don't quite remember how local file scheme works

Reply:
> please verify it locally first. I don't quite remember how local file scheme works

For local
```
C02D45VVMD6T:spark yi.zhu$ ./bin/spark-sql -f /Users/yi.zhu/Documents/test.sql
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
21/12/13 14:52:22 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
21/12/13 14:52:24 WARN HiveConf: HiveConf of name hive.stats.jdbc.timeout does not exist
21/12/13 14:52:24 WARN HiveConf: HiveConf of name hive.stats.retries.wait does not exist
21/12/13 14:52:26 WARN ObjectStore: Version information not found in metastore. hive.metastore.schema.verification is not enabled so recording the schema version 2.3.0
21/12/13 14:52:26 WARN ObjectStore: setMetaStoreSchemaVersion called but recording version is disabled: version = 2.3.0, comment = Set by MetaStore yi.zhu@10.12.189.45
Spark master: local[*], Application Id: local-1639378343821
1
Time taken: 3.869 seconds, Fetched 1 row(s)
2
Time taken: 0.052 seconds, Fetched 1 row(s)
3
Time taken: 0.042 seconds, Fetched 1 row(s)
C02D45VVMD6T:spark yi.zhu$
```


File in hdfs's /user/yi.zhu/test.sql
```
(base) [yi.zhu@ip-xx-xx-xx-xx ~]$ /usr/share/spark/bin/spark-sql --master local -f /user/yi.zhu/test.sql
21/12/13 14:56:10 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
21/12/13 14:56:10 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.
21/12/13 14:56:11 ERROR SparkSQLCLIDriver: Could not open input file for reading. (File file:/user/yi.zhu/test.sql does not exist)
```

```
(base) [yi.zhu@ip-10-130-21-35 ~]$ /usr/share/spark/bin/spark-sql --master local -f hdfs://R2/user/yi.zhu/test.sql
21/12/13 14:58:20 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
21/12/13 14:58:20 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.
1
Time taken: 2.072 seconds, Fetched 1 row(s)
Error in query:
mismatched input 'seelct' expecting {'(', 'SELECT', 'FROM', 'ADD', 'DESC', 'WITH', 'VALUES', 'CREATE', 'TABLE', 'INSERT', 'DELETE', 'DESCRIBE', 'EXPLAIN', 'SHOW', 'USE', 'DROP', 'ALTER', 'MAP', 'SET', 'RESET', 'START', 'COMMIT', 'ROLLBACK', 'REDUCE', 'REFRESH', 'CLEAR', 'CACHE', 'UNCACHE', 'DFS', 'TRUNCATE', 'ANALYZE', 'LIST', 'REVOKE', 'GRANT', 'LOCK', 'UNLOCK', 'MSCK', 'EXPORT', 'IMPORT', 'LOAD'}(line 2, pos 0)

== SQL ==

seelct 2
^^^

(base) [yi.zhu@ip-xx-xx-xx-xx ~]$ cat test.sql
select 1;
seelct 2;
select 3;
```